lib/game_client/ogs/territory_calculator.dart

```diff
@@ -0,0 +1,191 @@
+
+  // Count captured stones from manual removal
+  for (final removedPoint in removedStones) {
+    final removedStoneColor = boardState[removedPoint];
+    if (removedStoneColor == wq.Color.black) {
+      whiteScore += 1.0; // White gets a point for capturing black
+    } else if (removedStoneColor == wq.Color.white) {
+      blackScore += 1.0; // Black gets a point for capturing white
+    }
+  }
+
+  for (int r = 0; r < boardSize; r++) {
+    for (int c = 0; c < boardSize; c++) {
+      final point = (r, c);
+      final stoneColor = boardState[point];
+      final territoryColor = ownership[r][c];
+
+      // Count territory
+      if (stoneColor == null || removedStones.contains(point)) {
+        if (territoryColor == wq.Color.black) {
+          blackScore += 1.0;
+        } else if (territoryColor == wq.Color.white) {
+          whiteScore += 1.0;
+        }
+      }
+    }
+  }
+
+  // Add komi for white
+  whiteScore += komi;
+
+  // Determine winner and score lead
+  final scoreLead = (blackScore - whiteScore).abs();
+  final winner = blackScore > whiteScore ? wq.Color.black : wq.Color.white;
+
+  return CountingResult(
+    winner: winner,
+    scoreLead: scoreLead,
+    ownership: ownership,
+  );
+}
+
+/// Determine territory ownership for an empty region using flood fill
+wq.Color? _calculatePointTerritory(
+  BoardState boardState,
+  wq.Point startPoint,
+  List<List<bool>> visited,
+  List<wq.Point> removedStones,
+  int boardSize,
+) {
+  final queue = <wq.Point>[startPoint];
+  final surroundingColors = <wq.Color>{};
+
+  while (queue.isNotEmpty) {
+    final point = queue.removeAt(0);
+    final (r, c) = point;
+
+    if (r < 0 || r >= boardSize || c < 0 || c >= boardSize || visited[r][c]) {
+      continue;
+    }
+
+    final stoneColor = boardState[point];
+
+    if (stoneColor != null && !removedStones.contains(point)) {
+      // This is a living stone, record its color
+      surroundingColors.add(stoneColor);
+      continue;
+    }
+
+    // This is an empty point or removed stone
+    visited[r][c] = true;
+
+    // Add neighbors to queue
+    final neighbors = [(r - 1, c), (r + 1, c), (r, c - 1), (r, c + 1)];
+
+    for (final neighbor in neighbors) {
+      queue.add(neighbor);
+    }
+  }
+
+  // Determine ownership based on surrounding colors
+  if (surroundingColors.length == 1) {
+    // Territory is surrounded by only one color
+    return surroundingColors.first;
+  } else {
+    // Territory is contested or neutral
+    return null;
+  }
+}
+
+/// Fill a territory region with the specified color
+void _fillTerritoryRegion(
+  wq.Point startPoint,
+  wq.Color color,
+  List<List<wq.Color?>> ownership,
+  BoardState boardState,
+  List<wq.Point> removedStones,
+  int boardSize,
+) {
+  final queue = <wq.Point>[startPoint];
+  final visited = <wq.Point>{};
+
+  while (queue.isNotEmpty) {
+    final point = queue.removeAt(0);
+    final (r, c) = point;
+
+    if (r < 0 ||
```

Comment:
you can leverage [inside](https://github.com/ale64bit/WeiqiHub/blob/main/lib/board/board_state.dart#L179) for these and add the `visited` part to the condition.